test/jdk/java/net/httpclient/HttpClientClose.java

```diff
@@ -0,0 +1,345 @@
+
+    static String readBody(InputStream in) {
+        try {
+            return new String(in.readAllBytes(), StandardCharsets.UTF_8);
+        } catch (IOException io) {
+            throw new UncheckedIOException(io);
+        }
+    }
+
+    @Test(dataProvider = "positive")
+    void testConcurrent(String uriString) throws Exception {
+        out.printf("%n---- starting (%s) ----%n", uriString);
+        ExecutorService readerService = Executors.newCachedThreadPool();
+        Throwable failed = null;
+        try (HttpClient client = HttpClient.newBuilder()
+                .proxy(NO_PROXY)
+                .followRedirects(Redirect.ALWAYS)
+                .sslContext(sslContext)
+                .build()) {
+            TRACKER.track(client);
+
+            List<CompletableFuture<String>> bodies = new ArrayList<>();
+            for (int i = 0; i < ITERATIONS; i++) {
+                URI uri = URI.create(uriString + "/concurrent/iteration-" + i);
+                HttpRequest request = HttpRequest.newBuilder(uri)
+                        .header("X-uuid", "uuid-" + requestCounter.incrementAndGet())
+                        .build();
+                out.printf("Iteration %d request: %s%n", i, request.uri());
+                CompletableFuture<HttpResponse<InputStream>> responseCF;
+                CompletableFuture<String> bodyCF;
+                final int si = i;
+                responseCF = client.sendAsync(request, BodyHandlers.ofInputStream())
+                        .thenApply((response) -> {
+                            out.println(si + ":  Got response: " + response);
+                            assertEquals(response.statusCode(), 200);
+                            return response;
+                        });
+                bodyCF = responseCF.thenApplyAsync(HttpResponse::body, readerService)
+                        .thenApply(HttpClientClose::readBody)
+                        .thenApply((s) -> {
+                            assertEquals(s, MESSAGE);
+                            return s;
+                        });
+                long sleep = RANDOM.nextLong(5);
+                if (sleep > 0) {
+                    out.printf("%d: sleeping %d ms%n", i, sleep);
+                    Thread.sleep(sleep);
+                }
+                var cf = bodyCF;
+                bodies.add(cf);
+            }
+            CompletableFuture.allOf(bodies.toArray(new CompletableFuture<?>[0])).get();
+        } catch (Throwable throwable) {
+            failed = throwable;
+        } finally {
+            failed = cleanup(readerService, failed);
+        }
+        if (failed instanceof Exception ex) throw ex;
+        if (failed instanceof Error e) throw e;
+    }
+
+    static Throwable cleanup(ExecutorService readerService, Throwable failed) {
+        try {
+            readerService.shutdown();
+            readerService.awaitTermination(2000, TimeUnit.MILLISECONDS);
```

Comment:
Same comment about the timeout value.

Reply:
Same answer: there should be no operation in progress at this point - since we have waited for all the completable futures to complete, and we're not relying on the GC to garbage things.